crates/reth/evm/src/utils.rs

```diff
@@ -22,37 +23,52 @@ pub(crate) fn wei_to_sats(wei: U256) -> (U256, U256) {
     wei.div_rem(WEI_PER_SAT)
 }
 
-/// Tuple of executed transaction and receipt
-pub(crate) type TxReceiptPair<'a> = (&'a TransactionSigned, &'a Receipt);
-
-/// Collects withdrawal intents from bridge-out events by matching
-/// executed transactions (for txid) and receipts.
-/// Returns a vector of [`WithdrawalIntent`]s.
+/// Extracts withdrawal intents from bridge-out events in transaction receipts.
+/// Returns an iterator of [`WithdrawalIntent`]s.
 ///
 /// # Note
 ///
-/// A [`Descriptor`], if invalid does not create an [`WithdrawalIntent`].
-pub fn collect_withdrawal_intents<'a, I>(
-    tx_receipt_pairs: I,
-) -> impl Iterator<Item = WithdrawalIntent> + 'a
-where
-    I: Iterator<Item = TxReceiptPair<'a>> + 'a,
-{
-    tx_receipt_pairs.flat_map(|(tx, receipt)| {
-        let txid = Buf32((*tx.hash()).into());
-        receipt.logs.iter().filter_map(move |log| {
-            if log.address != BRIDGEOUT_PRECOMPILE_ADDRESS {
-                return None;
-            }
-
-            let event = WithdrawalIntentEvent::decode_log(log).ok()?;
-            let destination = Descriptor::from_bytes(&event.destination).ok()?;
-
-            Some(WithdrawalIntent {
-                amt: event.amount,
-                destination,
-                withdrawal_txid: txid,
+/// A [`Descriptor`], if invalid does not create a [`WithdrawalIntent`].
+pub fn withdrawal_intents<'a>(
+    transactions: &'a [TransactionSigned],
+    receipts: &'a [Receipt],
+) -> impl Iterator<Item = WithdrawalIntent> + 'a {
+    assert_eq!(
+        transactions.len(),
+        receipts.len(),
+        "transactions and receipts must have the same length"
+    );
```

Comment:
Either make sure to note this panic condition in the doc comment (`# Panics`) or make the return type be a result.